Implement 2D Scene ScaleMode

Yanrishatum opened the pull request:
* Closes #524, closes #563.
* Implements 8 different scaling modes for `s2d.Scene`.
* Deprecates `zoom`. Now can be done with `ScaleMode.Zoom(v)`
* Deprecates `setFixedSize`. Now can be done with `ScaleMode.Stretch(w, h)`.
* Removes size checks each frame and instead hooks to window resize events.

Now onto modes:
* `Fill` - it's the default and just resizes Scene to fill engine size.
* `Stretch` - basically good old ugly `setFixedSize`.
* `KeepAspect` - scales Scene with preservation of aspect-ratio. Can anchor on top-left and center.
* `KeepPixelSize` - same as above, but uses integer scaling to ensure square pixels.
* `Exact` - just forces width/height/zoom/centering disregarding adaptation.
* `Zoom` - Same as old `zoom` variable.
* `EnsureArea` - works same as KeepAspect, but instead of having fixed width and height - adaptas it to fill the screen. For example window 800x600 and `EnsureArea(400, 250)` would be upscaled to 2x and cover 800x500, but `height` will be `600`.
* `EnsurePixelArea` - same as above but integer scaling.

There is a particular issue with `drawTile` however - it checks for Scene bounds, not "is it visible on screen". It's logical but may be unexpected for some. We may keep it as is (`scaleMode` docs mention this behavior), force-clip Scene during rendering (`setRenderZone`) or rework how clipper in `drawTile` operates to check window-bounds instead of Scene bounds.

trethaller commented:
@ncannasse I've tested this with the UI scale mode in Northgard and didn't notice any issue. I'm happy to merge it if you agree

ncannasse commented:
Thanks for the PR and sorry for the review delay.

I'm having some difficulty when I put myself as a user to understand the expected outcome of each scale mode. That's maybe a question of vocabulary or documentation, but let's discuss it before merging.

Fill : I would instead maybe name it as Expand or something, as it enlarge the actual virtual area 

Stretch : ok

KeepAspect : I'm having some trouble with the meaning of width/height in this context. We give an aspect ratio Float (width/height) but then what if the screen does not have the same ratio as the specified one ? In that case when centered we either match the height (thus cutting/expending the left/right borders) or we match the width (thus cutting/expending top/left area).

So I think this should be actually two modes:   KeepWidth( width : Int, centered : Bool ) and KeepHeight( height : Int, centered : Bool )

KeepAspectPixel : I would remove this entirely, as it's very difficult to actually match the good pixel ratio. One way to implement it would be to give the following:
   Pixel( ?minWidth : Int, ?minHeight : Int, ?maxWidth : Int, ?maxHeight : Int );
And the algorithm would try to find the best fixed-size zoom that matches these constraints (if not null) while minimizing constraint breakage if some weird aspect ratio occurs.

Exact : I'm not sure at all what this is doing, it is a stretch with a zoom ?

Zoom : ok. but why would it need to resize "Engine" ? (should actually be named as "screen background buffer")

EnsureArea: from the naming it seems that it can scale down but not up ? and why would it need to resize backgound buffer then?

I guess it's quite difficult to come up with some well described schemes, as we have a scene with no fixed sized that we want to map to a screen with no fixed size either.

There is for example the Unity options for it:
https://docs.unity3d.com/Manual/script-CanvasScaler.html

Actually, once thing that seems important to me, and maybe the source of confusion, is that I think (0,s2d.width)/(0,s2d.height) should always be the _visible_ part of the scene on screen. 

If OTOH we want some kind of fixed size virtual scene which some parts would be actually hidden from screen, I think this would require more discussion about how we represent such area in terms of s2d API.

Yanrishatum commented:
I admit, my documentation wording and clarity could be better.

Fill: I don't think Expand would convey the function better. Because it adapts virtual size to window size. Expand by meaning would make user assume that it would grow, but will it actually shrink back? Fill on other hand clearly conveys that virtual size will fill entire window size.

KeepAspect: Width/height is here to define the internal virtual size. And when rendering, it will be stretched maintaining aspect ratio of width/height, and ensuring the whole area is visible at all times. It's basically letterboxing mode, so with 4:3 window - 16:9 virtual size will use up all horizontal space and result in letterboxing on top/bottom (or just bottom if centering is disabled). The virtual size remains constant at all times, and what you propose with KeepWidth/KeepHeight is more akin to EnsureArea.

KeepPixelSize: I'm not sure why would that break anything. This mode is basically KeepAspect, but upscale allowed only in integer steps (1x, 2x, 3x, 4x, 5x, etc), this mode is targeted towards pixel-art games which use low internal resolution (e.g. something like 320x240) but need the pixels remain square at all times. With non-integer scaling pixels are not guaranteed to be square. E.g. one pixel is 2x2, but pixel next to it 3x2.

Exact: This is "manual" mode basically. Constant virtual size and visual scale, as well as control on whether center the viewport or not. 

Zoom: Because it's how old `zoom` variable worked. :) 

EnsureArea: I couldn't come up with better name, but as with KeepAspect, it will upscale/downscale the rendered scene, but unlike KeepAspect the virtual size is not constant, but will guaranteed to be minimum of the specified size (e.g. ensured minimum virtual size). 320x200 scene on 800x600 screen would be upscaled 2.5 times. With KeepAspect, it would mean that there would be a letterboxing on top and bottom, because `200 * 2.5 = 500` with 100 pixels out. EnsureArea will include those remaining pixels in the virtual size, resulting in 320x240 virtual size. And it resizes background buffer because for 800x601, there's 1 pixel leftover and uncertain whether to use 320x240 or 320x241, and given that `zoom` solution was to upscale buffer - I did the same here too.

KeepWidth/KeepHeight: As I understood, that would be EnsureArea, but taking into consideration only width or height, which can be achieved with `EnsureArea(width, 0)` and `EnsureArea(height, 0)`

As for Unity:
* Constant Pixel Size = Fill
* Scale With Screen Size = KeepAspect
* Constant Physical Size - no analogue, because scaling mode does not work with DPI.

I suppose I need to work better on documentation explaining each mode? :)

ncannasse commented:
> I admit, my documentation wording and clarity could be better.
> 
> Fill: I don't think Expand would convey the function better. Because it adapts virtual size to window size. Expand by meaning would make user assume that it would grow, but will it actually shrink back? Fill on other hand clearly conveys that virtual size will fill entire window size.

What about Resize then ?

> KeepAspect: Width/height is here to define the internal virtual size. And when rendering, it will be stretched maintaining aspect ratio of width/height

I my mind, "stretched" != "maintaining aspect ratio"

>, and ensuring the whole area is visible at all times. It's basically letterboxing mode, so with 4:3 window - 16:9 virtual size will use up all horizontal space and result in letterboxing on top/bottom (or just bottom if centering is disabled). The virtual size remains constant at all times, and what you propose with KeepWidth/KeepHeight is more akin to EnsureArea.

So what about naming it LetterBox(width,height) ? I would either drop the centerViewPort entirely or make two enums that tells how to center (the two being optional and default to Center by default) :   LetterBox(width,height,Left|Right|Center,Top|Bottom|Center)

> KeepPixelSize: I'm not sure why would that break anything. This mode is basically KeepAspect, but upscale allowed only in integer steps (1x, 2x, 3x, 4x, 5x, etc), this mode is targeted towards pixel-art games which use low internal resolution (e.g. something like 320x240) but need the pixels remain square at all times. With non-integer scaling pixels are not guaranteed to be square. E.g. one pixel is 2x2, but pixel next to it 3x2.

I see, I still think this should be an option to LetterBox, please change accordingly.

> Exact: This is "manual" mode basically. Constant virtual size and visual scale, as well as control on whether center the viewport or not.

I see. Maybe "Fixed" is better naming then ? And my remarks about centering of LetterBox should be the same here.
 
> Zoom: Because it's how old `zoom` variable worked. :)

Let's slightly break backward compatibility here then : no resize of background buffer ;)
Makes me wonder if we should not change the s2d width/height to Float as we did for most other Ints.
 
> EnsureArea: I couldn't come up with better name, but as with KeepAspect, it will upscale/downscale the rendered scene, but unlike KeepAspect the virtual size is not constant, but will guaranteed to be minimum of the specified size (e.g. ensured minimum virtual size). 320x200 scene on 800x600 screen would be upscaled 2.5 times. With KeepAspect, it would mean that there would be a letterboxing on top and bottom, because `200 * 2.5 = 500` with 100 pixels out. EnsureArea will include those remaining pixels in the virtual size, resulting in 320x240 virtual size. And it resizes background buffer because for 800x601, there's 1 pixel leftover and uncertain whether to use 320x240 or 320x241, and given that `zoom` solution was to upscale buffer - I did the same here too.
> 
> KeepWidth/KeepHeight: As I understood, that would be EnsureArea, but taking into consideration only width or height, which can be achieved with `EnsureArea(width, 0)` and `EnsureArea(height, 0)`

I agree, but I think maybe it should be AutoZoom, and have the width/height optional to reflect the fact that they can be set to 0 to perform auto scaling of one of the dimensions only
 
> As for Unity:
> 
> * Constant Pixel Size = Fill
> * Scale With Screen Size = KeepAspect
> * Constant Physical Size - no analogue, because scaling mode does not work with DPI.
> 
> I suppose I need to work better on documentation explaining each mode? :)

Yes, let's first agree on terminology/API, then please write a bit more documentation for each case.

Yanrishatum commented:
* Renamed Fill -> Resize
* Renamed `KeepAspect` -> `LetterBox`
* Renamed `EnsureSize` -> `AutoZoom`
* Renamed `Exact` -> `Fixed`
* Unified integer scaling modes with non-integer by adding bool argument to `LetterBox` and `AutoZoom`.
* Added viewport align anchoring for `LetterBox` and `AutoZoom`
* Removed backbuffer resize.
* Added sample showcasing scaling modes.
* `posChanged` is set only if Scene size was changed (won't trigger for constant scene size modes when resizing the window)
* Reworked docs. Hope it's better now.

ncannasse commented:
Thanks for taking the time to update based on the feedback :)
